_posts/2018-09-07-redis-ruby-bloom-filter.md

```diff
@@ -0,0 +1,355 @@
+
+
+### Initial bloom filter implementation
+
+Initially, we used the C-implementation from [bloomfilter-rb](https://github.com/igrigorik/bloomfilter-rb) by building a bloom filter in-memory and uploading it to Amazon S3.
+
+There were several issues with this approach, the two most important ones being:
+
+- concurrency; sending multiple campaigns at the same time overrides the filter
+- slow updates / restricted to bulk updates only; fetching files from S3 is not very fast and updating the filter for one-off sends is expensive and not doable
+
+With the re-design, we need a solution that will solve these problems.
+
+
+### Bloom filter implementation
+
+Our bloom filter will have as a dependency our `User` model. Let's say the `User` model has three attributes: `id`, `max_contacts` and `billing_cycle_started_at`:
+
+```ruby
+User = Struct.new(:id, :max_contacts, :billing_cycle_started_at)
+user = User.new(1, 500, Time.new(2018, 8, 01, 10, 0, 0, 0))
+```
+
+Here is our bloom filter implementation:
+
+```ruby
+require 'zlib'
+
+class BloomFilter
+
+  # http://www.igvita.com/2008/12/27/scalable-datasets-bloom-filters-in-ruby/
+  # 10 bits for 1% error approximation
+  # ~5 bits per 10 fold reduction in error approximation
+  BITS_PER_ERROR_RATE = {
+    1    => 10,
+    0.1  => 15,
+    0.01 => 20
+  }
+  HASH_FUNCTIONS_COEFICIENT = 0.7 # Math.log(2)
+
+  attr_reader :error_rate
+
+  def initialize(user, error_rate: )
+    @user = user
+    @error_rate = error_rate
+  end
+
+  def indexes_for(key)
+    hash_functions.times.map { |i| Zlib.crc32("#{key.to_s.strip.downcase}:#{i+seed}") % size }
+  end
+
+  def hash_functions
+    @hash_functions ||= (bits * HASH_FUNCTIONS_COEFICIENT).ceil.to_i
+  end
+
+  def seed
+    @seed ||= since.to_i
+  end
+
+  def since
+    @since ||= @user.billing_cycle_started_at
+  end
+
+  def size
+    @size ||= bits * @user.max_contacts
+  end
+
+  def bits
+    @bits ||= BITS_PER_ERROR_RATE.fetch(error_rate)
+  end
+
+  def fingerprint
+    @fingerprint ||= [@user.id, @user.max_contacts, seed].join('.')
+  end
+end
+```
+
+The most important part of the bloom filter is the method that generates the indexes for a given key, in our case it is named `indexes_for(key)`.
+
+Here's an example usage:
+
+```ruby
+bloom_filter = BloomFilter.new(user, error_rate: 1)
+
+bloom_filter.indexes_for('user1@example.com')
+# [2872, 110, 3108, 2498, 4409, 751, 2861]
+
+bloom_filter.indexes_for('user2@example.com')
+# [3992, 2262, 1788, 1970, 3185, 4135, 4957]
+```
+
+As a hashing function we use [CRC32](https://en.wikipedia.org/wiki/Cyclic_redundancy_check) with a custom seed per user that is the `billing_cycle_started_at` and a number of hashing functions based on the error rate (in this example we use an error rate of 1%).
```

Comment:
I'm not an expert in security by any means, but is it possible knowing some of this stuff would give hackers information how to defeat it?